src/net/protocol_agnostic/stream.rs

```diff
@@ -36,6 +44,124 @@ impl PaStream {
         }
     }
 
+    pub fn rendezvous_connect<C>(
```

Comment:
That is quite complex function with a lot of logic and branching - would you care to decompose it?